src/c++/perf_analyzer/test_request_rate_manager.cc

```diff
@@ -387,4 +405,65 @@ TEST_CASE("request_rate_sequence")
   trrm.TestSequences();
 }
 
+/// Check custom json data to ensure that it is processed correctly
+///
+TEST_CASE("custom_json_data")
+{
+  PerfAnalyzerParameters params;
+
+  SUBCASE("non sequence model")
+  {
+    bool is_seq_model = false;
+    bool using_mock_infer = false;
+    TestRequestRateManager trrm{params, is_seq_model, using_mock_infer};
+    std::unique_ptr<MockDataLoader> mdl = std::make_unique<MockDataLoader>();
+    MockModelParser mmp{false};
+
+    auto thread_status = std::make_shared<TestRequestRateManager::ThreadStat>();
+    thread_status->status_ = cb::Error::Success;
+    thread_status->cb_status_ = cb::Error::Success;
+
+    auto thread_config =
+        std::make_shared<TestRequestRateManager::ThreadConfig>(0, 0);
+    thread_config->non_sequence_data_step_id_ = 0;
+
+
+    std::string ss = R"(
+  {
+      "data" : [
+        {
+          "INPUT0" : [ 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1 ],
+          "INPUT1" : [ 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1 ]
+        },
+        {
+          "INPUT0" : [ 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1 ],
+          "INPUT1" : [ 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1 ]
+        },
+        {
+          "INPUT0" : [ 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1 ],
+          "INPUT1" : [ 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1 ]
+        },
+        {
+          "INPUT0" : [ 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1 ],
+          "INPUT1" : [ 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1 ]
+        }
+      ]
+  }
+  )";
+
+    mdl->ReadDataFromStr(mmp.Inputs(), mmp.Outputs(), ss);
+    trrm.data_loader_ = std::move(mdl);
+    trrm.using_json_data_ = true;
+    CHECK(trrm.on_sequence_model_ == false);
+    bool expect_ok = true;
+    trrm.execute_ = true;
+    trrm.schedule_.push_back(std::chrono::nanoseconds(1));
+    std::thread infer_thread(
+        &TestRequestRateManager::Infer, &trrm, thread_status, thread_config);
+    early_exit = true;
+    infer_thread.join();
+    CHECK(thread_status->status_.IsOk() == expect_ok);
+    CHECK(thread_config->non_sequence_data_step_id_ == 4);
```

Comment:
What are we actually testing here? Is non_sequence_data_step_id_ the main/only side-effect from using json data?

Reply:
it is the only side effect that was easily observable without digging into more layers of the data loader, which seemed out of scope for this work.